commands/upgrade_cluster.go

```diff
@@ -0,0 +1,344 @@
+			subtext = fmt.Sprintf("We couldn't find a cluster with the ID '%s' via API endpoint %s.", args.clusterID, args.apiEndpoint)
+		case IsCommandAbortedError(err):
+			headline = "Not upgrading."
+		default:
+			headline = err.Error()
+		}
+
+		// Print error output
+		fmt.Println(color.RedString(headline))
+		if subtext != "" {
+			fmt.Println(subtext)
+		}
+		os.Exit(1)
+	}
+
+	fmt.Println(color.GreenString("Starting to upgrade cluster '%s' to release version %s",
+		color.CyanString(result.clusterID),
+		color.CyanString(result.versionAfter)))
+}
+
+// upgradeCluster performs our actual function. It usually creates an API client,
+// configures it, configures an API request and performs it.
+func upgradeCluster(args upgradeClusterArguments) (upgradeClusterResult, error) {
+	result := upgradeClusterResult{}
+
+	// fetch current cluster details
+	details, detailsErr := getClusterDetails(args.clusterID, args.authToken, args.apiEndpoint)
+	if detailsErr != nil {
+		return result, microerror.Mask(detailsErr)
+	}
+
+	listReleasesArgs := listReleasesArguments{
+		apiEndpoint: args.apiEndpoint,
+		token:       args.authToken,
+	}
+	releasesResult, releasesErr := listReleases(listReleasesArgs)
+	if releasesErr != nil {
+		return result, microerror.Mask(releasesErr)
+	}
+	releaseVersions := []string{}
+	for _, r := range releasesResult.releases {
+		releaseVersions = append(releaseVersions, r.Version)
+	}
+
+	// define the target version to upgrade to
+	targetVersion := successorReleaseVersion(details.ReleaseVersion, releaseVersions)
+	if targetVersion == "" {
+		return result, microerror.Mask(noUpgradeAvailableError)
+	}
+
+	var targetRelease gsclientgen.V4ReleaseListItem
+	for _, rel := range releasesResult.releases {
+		if rel.Version == targetVersion {
+			targetRelease = rel
+		}
+	}
+
+	// Show some details independent of confirmation
+	if !targetRelease.Active {
+		fmt.Printf("Cluster '%s' will be upgraded from version %s to %s, which is not an active release.\n",
+			color.CyanString(args.clusterID),
+			color.CyanString(details.ReleaseVersion),
+			color.CyanString(targetVersion))
+		fmt.Printf("This might fail depending on your permissions.\n")
+	} else {
+		fmt.Printf("Cluster '%s' will be upgraded from version %s to %s.\n",
+			color.CyanString(args.clusterID),
+			color.CyanString(details.ReleaseVersion),
+			color.CyanString(targetVersion))
+	}
+
+	// Details output and confirmation
+	if !args.force {
+
+		fmt.Println("")
+		fmt.Println("Changelog:")
+		fmt.Println("")
+
+		for _, change := range targetRelease.Changelog {
+			fmt.Printf("    - %s: %s\n", change.Component, change.Description)
+		}
+
+		fmt.Println("")
+		fmt.Println("NOTE: Upgrading may impact your running workloads and will make the cluster's")
+		fmt.Println("Kubernetes API unavailable temporarily. Before upgrading, please acknowledge the")
+		fmt.Println("details described in")
+		fmt.Println("")
+		fmt.Printf("    %s\n", upgradeDocsURL)
+		fmt.Println("")
```

Comment:
Why not printing the changelog and note when args.force?

Reply:
The idea was that if the user uses -f/--force, that information couldn't really be used anyway.

Now that I'm thinking about it, it might still be that the user wants to scroll back and find out what really happened. And -f/--force is useful for automations, where that output would still be stored somewhere. OK, I'll change that.

Reply:
Yeah, when you don't care about the output `> /dev/null`.

Reply:
Changed